File: 210. Course Schedule II.js
var findOrder = function(numCourses, prerequisites) {
    preq=new Map
    for ([a,b] of prerequisites){
        preq.has(a)? preq.get(a).push(b) : preq.set(a,[b])                
    }
    seen=new Set
    loop=new Set
    let flag=true
    
    function dfs(i) {
        if (loop.has(i)) {
            flag=false           
            return
        }

        if (seen.has(i)) return        
        seen.add(i)        
        loop.add(i)      
        
        let pre=preq.get(i)        
        if (pre)  {for (c of pre) dfs(c)}
        loop.delete(i)              
        path.push(i) 
        return 
        
    
    }    
    
    path=[]
    for(let i=0;i<numCourses;i++){        
       dfs(i)        
    }

return (flag ? path :[])
};

File: 210. Course Schedule II.py
class Solution:
    def findOrder(self, numCourses: int, prerequisites: List[List[int]]) -> List[int]:
        preq=defaultdict(list)
        for c,p in prerequisites:
            preq[c].append(p)
        seen=set()
        res=[]
        cycle=set()
        isCycle=False
        def dfs(course):
            nonlocal isCycle
            if isCycle==True or course in seen:
                return 
            if course in cycle :
                isCycle=True
                return
            cycle.add(course)
            for c in preq[course]:
                dfs(c)
            seen.add(course)
            res.append(course)
            cycle.remove(course)
        for c in range(numCourses):
            dfs(c)        
            if isCycle:
                return []
        return res


